File: docs/code-examples/box3d_batch.cpp
// Log a batch of oriented bounding boxes.

#include <rerun.hpp>

int main() {
    auto rec = rerun::RecordingStream("rerun_example_box3d_batch");
    rec.connect("127.0.0.1:9876").throw_on_failure();

    rec.log(
        "batch",
        rerun::Boxes3D::from_centers_and_half_sizes(
            {{2.0f, 0.0f, 0.0f}, {-2.0f, 0.0f, 0.0f}, {0.0f, 0.0f, 2.0f}},
            {{2.0f, 2.0f, 1.0f}, {1.0f, 1.0f, 0.5f}, {2.0f, 0.5f, 1.0f}}
        )
            .with_rotations({
                rerun::datatypes::Quaternion::IDENTITY,
                // 45 degrees around Z
                rerun::datatypes::Quaternion(0.0f, 0.0f, 0.382683f, 0.923880f),
                rerun::datatypes::RotationAxisAngle(
                    {0.0f, 1.0f, 0.0f},
                    rerun::datatypes::Angle::degrees(30.0f)
                ),
            })
            .with_radii({0.025f})
            .with_colors({
                rerun::datatypes::Rgba32(255, 0, 0),
                rerun::datatypes::Rgba32(0, 255, 0),
                rerun::datatypes::Rgba32(0, 0, 255),
            })
            .with_labels({"red", "green", "blue"})
    );
}

File: docs/code-examples/box3d_batch.rs
//! Log a batch of oriented bounding boxes.
use rerun::{
    components::Color, Angle, Boxes3D, Quaternion, RecordingStreamBuilder, Rotation3D,
    RotationAxisAngle,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let (rec, storage) = RecordingStreamBuilder::new("rerun_example_box3d_batch").memory()?;

    rec.log(
        "batch",
        &Boxes3D::from_centers_and_half_sizes(
            [(2.0, 0.0, 0.0), (-2.0, 0.0, 0.0), (0.0, 0.0, 2.0)],
            [(2.0, 2.0, 1.0), (1.0, 1.0, 0.5), (2.0, 0.5, 1.0)],
        )
        .with_rotations([
            Rotation3D::IDENTITY,
            Quaternion::from_xyzw([0.0, 0.0, 0.382683, 0.923880]).into(), // 45 degrees around Z
            RotationAxisAngle::new((0.0, 1.0, 0.0), Angle::Degrees(30.0)).into(),
        ])
        .with_radii([0.025])
        .with_colors([
            Color::from_rgb(255, 0, 0),
            Color::from_rgb(0, 255, 0),
            Color::from_rgb(0, 0, 255),
        ])
        .with_labels(["red", "green", "blue"]),
    )?;

    rerun::native_viewer::show(storage.take())?;
    Ok(())
}
